Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorState.cs

```diff
@@ -54,34 +56,94 @@ internal struct InputActionsEditorState
+                : Selectors.GetActionCount(selectedActionMap) > 0 ? 0 : -1;
+
+            // Attempt to preserve binding selection by GUID, otherwise select first or none
+            m_selectedBindingIndex = -1;
+            if (m_selectedActionMapIndex >= 0)
+            {
+                var otherSelectedBinding = Selectors.GetSelectedBinding(other);
+                if (otherSelectedBinding != null)
+                {
+                    var otherSelectedBindingId =
+                        InputActionSerializationHelpers.GetId(otherSelectedBinding.Value.wrappedProperty);
+                    var binding = Selectors.GetBindingForId(asset, otherSelectedBindingId.ToString(), out _);
+                    if (binding != null)
+                        m_selectedBindingIndex = binding.GetIndexOfArrayElement();
+                }
+            }
+
+            // Sanity check selection type and override any previous selection if not valid given indices
+            // since we have remapped GUIDs to selection indices for another asset (SerializedObject)
+            if (other.m_selectionType == SelectionType.Binding && m_selectedBindingIndex < 0)
+                m_selectionType = SelectionType.Action;
+            else
+                m_selectionType = other.m_selectionType;
+
             m_selectedControlSchemeIndex = other.m_selectedControlSchemeIndex;
             m_selectedDeviceRequirementIndex = other.m_selectedDeviceRequirementIndex;
 
-            // Selected ControlScheme index is serialized but we have to recreated actual object after domain reload
-            if (m_selectedControlSchemeIndex != -1)
+            // Selected ControlScheme index is serialized but we have to recreated actual object after domain reload.
+            // In case asset is different from from others asset the index might not even be valid range so we need
+            // to reattempt to preserve selection but range adapt.
+            // Note that control schemes and device requirements currently lack any GUID/ID to be uniquely identified.
+            var controlSchemesArrayProperty = serializedObject.FindProperty(nameof(InputActionAsset.m_ControlSchemes));
+            if (m_selectedControlSchemeIndex >= 0 && controlSchemesArrayProperty.arraySize > 0)
             {
-                var controlSchemeSerializedProperty = serializedObject
-                    .FindProperty(nameof(InputActionAsset.m_ControlSchemes))
-                    .GetArrayElementAtIndex(m_selectedControlSchemeIndex);
-
-                m_ControlScheme = new InputControlScheme(controlSchemeSerializedProperty);
+                if (m_selectedControlSchemeIndex >= controlSchemesArrayProperty.arraySize)
+                    m_selectedControlSchemeIndex = 0;
+                m_ControlScheme = new InputControlScheme(
+                    controlSchemesArrayProperty.GetArrayElementAtIndex(other.m_selectedControlSchemeIndex));
+                // TODO Preserve device requirement index
```

Comment:
Incomplete TODO - not sure if this needs doing before landing. Seems no worse than it was